### ISSUE TITLE
Camera parameters

### ISSUE THREAD

**Issue by Ainaz99** [apple/ml-hypersim#24]
Hi,

I'm trying to render surface normals from the meshes in Blender, using the camera parameters provided. I scale the mesh using `meters_per_asset_unit`. I use `camera_keyframe_positions.hdf5` for the camera location and get the camera euler angles from `camera_keyframe_orientations.hdf5'` with `fov = pi/3 = 1.0471975511965976`. 
But my renderings do not match the color images from the dataset for only some specific scenes. Is there any other camera parameter I'm missing?
Here's an example for `ai_037_002`, `cam_00`, `frame.0000`:
![image](https://user-images.githubusercontent.com/69212889/118339474-2f341e80-b52e-11eb-9852-78ba2d4f025f.png)

Thanks for your help.

**Comment by mikeroberts3000** [comment]
Hi! Great question.

It sounds like you're doing everything right. I have noticed this problem occasionally also. It appears to affect a small handful of the V-Ray scenes, but I don't have a complete list. The problem occurs because there is occasionally an additional offset in the V-Ray scene file that gets applied to the camera position during photorealistic rendering, but doesn't get properly accounted for when we export the camera positions. I believe the relevant parameters are the "Horizontal Shift" and "Vertical Shift" parameters described in the [V-Ray documentation](https://docs.chaosgroup.com/display/VMAX/VRayPhysicalCamera). If I had known about these unpleasant parameters prior to rendering our images, I certainly would have explicitly hard-coded them to 0 for each scene.

Perhaps you can help with this issue. I'm going to assume that you have access to the source assets, and you have run the pre-processing phase of our pipeline. Otherwise, how could you do your own rendering in Blender? Anyway, there are several ways to proceed.

First, can you manually confirm that the scene you're experimenting with has non-zero values for these shift parameters in its exported vrscene file?

Second, given non-zero shift parameters from a vrscene file, and an (incorrectly) exported camera position and orientation from the Hypersim pipeline, can you compute the correct camera position? When using the correct camera position, your Blender renderings should match our ground truth images pixel-perfectly.
 
Third, do you have a complete list of scenes that are affected? It would be straightforward to parse every vrscene file and search for any non-zero shift parameters. See [this code](https://github.com/apple/ml-hypersim/blob/master/examples/01_marketplace_dataset/_vray_user_params.py#L75) for an example of how to programmatically access the cameras in a vrscene file.

**Comment by rpautrat** [comment]
Hi @mikeroberts3000,

I am also interested in multi-view applications based on Hypersim, and I stumbled upon the same issue as Ainaz99, Frank-Mic and rikba. It seems that a considerable amount of scenes are concerned with the shifted camera parameters and are thus unusable for multi-view applications.

From the reprojections tests that I have run, I think that non-zero tilt offsets are also involved, in addition to the shift offsets. Unfortunately I do not have access to the source assets and cannot run the steps that you suggested to investigate this issue. Did you have by chance time to investigate it and got fresh news about it?

If not, would it be possible to publicly expose the VRay tilt and shift parameters for each scene and cameras? I think that based on this information, I would be able to retrieve the correct camera position and orientation.

Overall I would be happy to contribute to the resolution of this problem, but I feel that some information coming from the source assets are necessary to achieve it, which I unfortunately do not have...

**Comment by mikeroberts3000** [comment]
Hi @rpautrat, I agree with your assessment that it is necessary to expose some additional information from the source assets. Once this additional information is exposed, it should be possible to derive the correct camera intrinsics for these scenes. I'll follow up with you offline and maybe we can work on this together.

**Comment by Ainaz99** [comment]
Hi @mikeroberts3000 @rpautrat ! Are there any updates on the correct camera parameters?

**Comment by mikeroberts3000** [comment]
Hi @Ainaz99, I'm happy to say that we're making solid progress. @rpautrat has been doing a bunch of great experiments to figure all of this out.

We identified four V-Ray parameters that can affect the camera intrinsics, and we have an accurate mathematical model of what each of these parameters does in isolation. Roughly speaking, each parameter translates or rotates the image plane in camera space. But we don't have a model of how the parameters interact with each other, and there are many possible conventions (translation then rotation, rotation then translation, different Euler angle conventions, etc). We reached out to Chaos Group, and we're waiting for them to tell us the exact order of operations.

Suppose we knew exactly how to transform the image plane as a function of our V-Ray parameters. Let's call this transformation T. We have sketched out a derivation for the non-standard projection matrix P, computed as a function of T, that correctly projects points from world space to image space. This projection matrix P can then be used as a drop-in replacement for the usual pinhole camera perspective projection matrix.

To summarize, we think we have a solid understanding of this issue, but we're waiting for Chaos Group to tell us exactly how to compute T based on the V-Ray parameters. If you're super motivated to get this issue resolved, and you don't want to wait for Chaos Group to get back to us, I'd be happy to send you the Jupyter notebooks that we've been using in our experiments. You don't need to install V-Ray to run our notebooks, and you can try to compute T from the V-Ray parameters with a brute force guess-and-check strategy.

**Comment by Ainaz99** [comment]
Thanks @mikeroberts3000, happy to hear you made progress in solving the issue. Do you have any estimate on how long it will take the Chaos Group to answer you?

**Comment by mikeroberts3000** [comment]
@Ainaz99 no estimate from Chaos Group. I'll definitely post here when we hear back from them. In the meantime, the invitation is still open for you to experiment with our experimental notebooks and attempt to compute the correct transformation 🤓

**Comment by liuzhy71** [comment]
> @Ainaz99 no estimate from Chaos Group. I'll definitely post here when we hear back from them. In the meantime, the invitation is still open for you to experiment with our notebooks and attempt to compute the correct transformation nerd_face
> 
> An alternative to the guess-and-check strategy would be to set up a simple scene, and infer the correct transformation from rendered observations. It would then be possible to collect a training set of (camera parameter, transformation) pairs, and fit a function to the training set.

How am I able to acquire the notebook for the camera operations? Can you send a copy of that notebook you mentioned before?

**Comment by mikeroberts3000** [comment]
Hi @liuzhy71, you're a total legend for having a look at this. I'll send you an email with all of our debugging notebooks.

**Comment by mikeroberts3000** [comment]
In case anyone else is interested, here is a diagram explaining what we think is going on.

<img width="583" alt="image001" src="https://user-images.githubusercontent.com/2341965/137516857-4337cac7-ede3-43f2-a8e0-c80888f12d13.png">

We think the image plane is being transformed somehow, and there seem to be four camera parameters that control the transformation. So, our goal is to determine the function _f_ that maps from the scalar camera parameters (p1,p2,p3,p4) to a transformation matrix T.

In our debugging notebook, we try to guess this function _f_ based on a code snippet that we got from Chaos Group. This code snippet is correct for some combinations of parameters, but not others. So we must be getting something wrong. In order to test the correctness of _f_ in our notebook, we compute a depth image using my own reference raycaster. In this test, we can control the outbound ray at each pixel. Using my reference raycaster, we want to obtain depth images that perfectly match the ones we obtain from V-Ray for all combinations of camera parameters. If we can do this, then we will know that we are implementing _f_ correctly.

To make progress, I think a promising approach would be the following. First, compute the correct transformation T, given a depth image rendered by V-Ray with a known set of camera parameters (p1,p2,p3,p4). The transformation can be recovered from the depth image by solving a convex optimization problem. Anyway, once this is working, it is straightforward to render a large collection of depth images with randomly chosen camera parameters, solving for T for each rendered image. After doing so, we will have a large collection of (camera parameters, transformation) pairs. Finally, it is straightforward to fit some kind of simple function (e.g., a neural network) to all of the example pairs. This learned function can then be queried later to output the correct transformation T for a new set of parameters (p1,p2,p3,p4). Once the transformation T is known, then it is straightforward to derive a modified projection matrix P (in terms of T) that projects world-space points to the correct image-space coordinates. This projection matrix can used as a drop-in replacement for the usual projection matrix in graphics and vision applications (e.g., multi-view stereo applications, rendering additional images that exactly match our pre-rendered Hypersim images, etc).

Of course doing all of this is an unpleasant hack. But so far, Chaos Group has been unable to mathematically characterize their own camera parameters, so we must resort to reverse-engineering their meaning.

If anyone is interested in having a look at this issue, comment here and I will send you our debugging notebooks.

**Comment by alexsax** [comment]
@mikeroberts3000 @liuzhy71 @Ainaz99 @rpautrat thank you all for the hard work on getting accurate camera parameters for these scenes.

Any update on when these will be released and we can use them? 😄

**Comment by jatentaki** [comment]
I'd also be curious if the fixes have been released somewhere :)

**Comment by mikeroberts3000** [comment]
Hi @Ainaz99 @alexsax @jatentaki @liuzhy71 @rpautrat, I have some good news. I just checked in some data and example code that resolves this issue.

In the [`contrib/mikeroberts3000`](https://github.com/apple/ml-hypersim/tree/master/contrib/mikeroberts3000) directory, I provide a modified perspective projection matrix for each scene that can be used as a drop-in replacement for the usual OpenGL perspective projection matrix, as well as example code for projecting world-space points into Hypersim images.

I apologize that it took so long to address this issue. It was especially challenging to debug because V-Ray's behavior wasn't well-documented, and I've been busy with other projects and holiday travel.

**Comment by Ainaz99** [comment]
Hi @mikeroberts3000 ,

Thanks for releasing the modified version of the camera parameters. 
I'm wondering if there is a way to directly calculate new camera positions and orientations which were originally saved as `camera_keyframe_positions.hdf5` and `camera_keyframe_orientations.hdf5` only using the released perspective projection matrices for the scenes? And if yes, how is the calculation done? Thank you!

**Comment by mikeroberts3000** [comment]
Hi @Ainaz99, this is possible, but there are some important technical details to be aware of. In particular, the modified "camera orientation" will have some extra transformations encoded in it, and it will not be a rotation matrix. As a result, any downstream code that intends to invert this matrix must take care to actually invert it, rather than merely transposing it.

To derive our modified camera orientation, consider the following equation that transforms a point in world-space `p_world` (expressed in homogeneous coordinates) into a point in homogeneous clip space `p_clip`,

```
p_clip = M_proj_modified * M_cam_from_world * p_world
```

where `M_proj_modified` is our modified projection matrix; and `M_cam_from_world` is a 4x4 matrix that encodes the camera position and orientation. We can express this transformation in terms of the standard OpenGL projection matrix `M_proj_canonical` as follows,

```
p_clip = M_proj_canonical * block_diag(M_cam_from_uv_canonical * M_cam_from_uv.I, 1) * M_cam_from_world * p_world
```

where `M_cam_from_uv` is a 3x3 matrix defined in our CSV file for each scene, and `M_cam_from_uv_canonical = diag([tan(fov_x/2.0), tan(fov_y/2.0), -1.0])`. Collecting the terms in the matrix `block_diag(...) * M_cam_from_world`, we see that we can define a modified camera orientation `R_world_from_cam_modified` that completely accounts for our non-standard camera parameters.

```
R_world_from_cam_modified = camera_orientation * (M_cam_from_uv_canonical * M_cam_from_uv.I).I
```

where `camera_orientation` is the camera orientation from our HDF5 file.

`R_world_from_cam_modified` can be used as a drop-in replacement for the camera orientations stored in our HDF5 files, and will completely account for all non-standard camera parameters when used in conjunction with a standard OpenGL projection matrix. In other words, we do not need to make any modifications to the camera positions stored in our HDF5 files. But remember that `R_world_from_cam_modified` is not a rotation matrix.

**Comment by Ainaz99** [comment]
Thank you @mikeroberts3000 for your explanation!
So if I want to calculate the `worldspace transformation matrix` for the camera as follows,

```
R_world_from_cam    =  camera_orientation
t_world_from_cam     =  camera_position.T
M_world_from_cam   =  matrix(block([[R_world_from_cam, t_world_from_cam], [matrix(zeros(3)), 1.0]]))
```

can I use the new `R_world_from_cam_modified`?
Thank you.

**Comment by Ainaz99** [comment]
Thank you @mikeroberts3000. So I want to use the new camera transformation matrix in Blender.
I'm attaching some images which show from left to right: the `RGB`, the `old rendering` (using `old M_world_from_cam`), the `new rendering` (using the `R_world_from_cam_modified`), and the RGB and new rendering overlaid. Although the new rendering looks better, there is still some difference which I assume it's due to the extra transformations that you mentioned. 
Do you know how I can take care of these extra transformations? Are there any other parameters I have to manually change for the camera?
![Screen Shot 1400-11-06 at 13 27 24](https://user-images.githubusercontent.com/69212889/151166818-ce7cc91e-ed0a-44d8-a4f5-0f5683c1fcf5.png)
![Screen Shot 1400-11-06 at 13 29 39](https://user-images.githubusercontent.com/69212889/151166830-4a640bae-2778-4cc3-9c92-4aff9bc5a3b9.png)
![Screen Shot 1400-11-06 at 13 30 16](https://user-images.githubusercontent.com/69212889/151166836-59bb1328-bf47-4d3c-9e0c-cd3d08c274c9.png)

**Comment by mikeroberts3000** [comment]
I haven't spent much time with Blender, so I'm not sure what exactly it is doing with the matrices you're specifying. Is Blender rendering images via a rasterization approach or a raycasting approach? How exactly is are you specifying these matrices to Blender? Are you specifying position and orientation in a combined 4x4 `M_world_from_cam_modified` matrix? Or are you specifying `R_world_from_cam_modified` and `t_world_from_cam` separately?

In [these](https://github.com/apple/ml-hypersim/blob/master/contrib/mikeroberts3000/jupyter/00_projecting_points_into_hypersim_images.ipynb) [notebooks](https://github.com/apple/ml-hypersim/blob/master/contrib/mikeroberts3000/jupyter/01_casting_rays_that_match_hypersim_images.ipynb), I show how to reproduce the Hypersim images exactly using a rasterization approach and a raycasting approach for this specific scene. It should be straightforward to find out what Blender is doing that is different to these notebooks by digging into the documentation. For what it's worth, in my local repository, I computed `R_world_from_cam_modified` according to the equation I posted above and I verified that it behaves as expected when used during rasterization and raycasting.

**Comment by mikeroberts3000** [comment]
Hi @Ainaz99, I'm just double-checking if you ever got this rendering functionality figured out in Blender.

**Comment by huntorochi** [comment]
@Ainaz99 @mikeroberts3000 
Hi,

I am working with the provided camera pose data from the Hypersim dataset, specifically the files:

``
camera_positions_hdf5_file    = os.path.join(camera_dir, "camera_keyframe_positions.hdf5")
camera_orientations_hdf5_file = os.path.join(camera_dir, "camera_keyframe_orientations.hdf5")

with h5py.File(camera_positions_hdf5_file,    "r") as f: 
    camera_positions = f["dataset"][:]
with h5py.File(camera_orientations_hdf5_file, "r") as f: 
    camera_orientations = f["dataset"][:]

camera_position_world = camera_positions[frame_id]
R_world_from_cam      = camera_orientations[frame_id]
``
From the above code, I understand that the dataset provides the camera's position (camera_position_world) and orientation (R_world_from_cam) in the 3ds Max coordinate system.

I need to convert this data into Blender's transform_matrix (4x4), considering the differences in the coordinate systems between 3ds Max and Blender. 

Could you clarify the exact steps required to transform the provided position and rotation data into Blender's coordinate system? If possible, an example of the transformation process or the correct transformation matrix would be very helpful.

Thank you for your assistance!

**Comment by mikeroberts3000** [comment]
@huntorochi You should figure this out yourself and post a self-contained code snippet here when you do. Start with the Hypersim scenes that don't have weird tilt-shift photography parameters.
